chain/client/src/stateless_validation/partial_witness/partial_witness_actor.rs

```diff
@@ -359,19 +358,34 @@ impl PartialWitnessActor {
             self.epoch_manager.as_ref(),
             &accesses,
             &signer,
-            &self.store,
+            self.runtime.store(),
         )? {
             return Ok(());
         }
         let key = accesses.chunk_production_key();
-        let contract_hashes = BTreeSet::from_iter(accesses.contracts().iter().cloned());
+        let contracts_cache = self.runtime.compiled_contract_cache();
+        let runtime_config = self
+            .runtime
+            .get_runtime_config(self.epoch_manager.get_epoch_protocol_version(&key.epoch_id)?)?;
+        let missing_contract_hashes = BTreeSet::from_iter(
+            accesses
+                .contracts()
+                .iter()
+                .filter(|hash| {
+                    !contracts_cache
+                        .has(&get_contract_cache_key(hash.0, &runtime_config.wasm_config))
```

Comment:
`has` does not check memory cache but only filesystem, so using memory cache may speed up the check.
see here: https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/runtime/near-vm-runner/src/near_vm_runner/runner.rs#L229-L239

Reply:
good point, implemented in https://github.com/near/nearcore/pull/12300/commits/e2f238d079e5d53542c15fa498fff65badd225d2